src/boot/efi/boot.c

```diff
@@ -2672,7 +2672,10 @@ EFI_STATUS efi_main(EFI_HANDLE image, EFI_SYSTEM_TABLE *sys_table) {
 
         export_variables(loaded_image, loaded_image_path, init_usec);
 
-        err = open_volume(loaded_image->DeviceHandle, &root_dir);
+        if (is_direct_boot(loaded_image->DeviceHandle))
+                err = qemu_open(&loaded_image->DeviceHandle, &root_dir);
+        else
+                err = open_volume(loaded_image->DeviceHandle, &root_dir);
```

Comment:
hmm, can we split this out into a function of its own?

I am pretty sure sooner or later we want other way to derive the root dir, for example via command line parameter, or via efi variable. If we pluralize how we find the root dir, we might as well split this now into its own function `discover_root_dir()` or so right now, to minimize later refactoring. two different paths is already enough to deserve a split-out.. efi_main() is huge anyway